File: Assets/Extreal/NGO/WebRTC/IdMapper.cs
using System;
using System.Collections.Generic;

namespace Extreal.NGO.WebRTC.Dev
{
    public class IdMapper
    {
        private readonly Dictionary<string, ulong> strToLongMapping = new Dictionary<string, ulong>();
        private readonly Dictionary<ulong, string> ulongToStrMapping = new Dictionary<ulong, string>();

        public void Add(string id)
        {
            var ulongId = Generate();
            strToLongMapping.Add(id, ulongId);
            ulongToStrMapping.Add(ulongId, id);
        }

        private ulong Generate()
        {
            var now = DateTimeOffset.UtcNow;
            var id = now.ToUnixTimeMilliseconds() + strToLongMapping.Count;
            return (ulong)id;
        }

        public bool Has(string id) => strToLongMapping.ContainsKey(id);

        public ulong Get(string id) => strToLongMapping[id];

        public bool Has(ulong id) => ulongToStrMapping.ContainsKey(id);

        public string Get(ulong id) => ulongToStrMapping[id];

        public void Remove(string id)
        {
            if (!Has(id))
            {
                return;
            }
            var ulongId = strToLongMapping[id];
            strToLongMapping.Remove(id);
            ulongToStrMapping.Remove(ulongId);
        }

        public void Clear()
        {
            strToLongMapping.Clear();
            ulongToStrMapping.Clear();
        }
    }
}

File: Assets/WebGLScripts/src/NGO/IdMapper.ts
class IdMapper {
    private readonly strToNumMapping: Map<string, number> = new Map();
    private readonly numToStrMapping: Map<number, string> = new Map();

    public add = (id: string) => {
        const numId = this.generate();
        this.strToNumMapping.set(id, numId);
        this.numToStrMapping.set(numId, id);
    };

    private generate = () => {
        const now = new Date();
        return now.getTime() + this.strToNumMapping.size;
    };

    public has = (id: string | number) => {
        return Number.isFinite(id) ? this.numToStrMapping.has(id as number) : this.strToNumMapping.has(id as string);
    };

    public get = (id: string | number) => {
        return Number.isFinite(id) ? this.numToStrMapping.get(id as number) : this.strToNumMapping.get(id as string);
    };

    public remove = (id: string) => {
        if (!this.has(id)) {
            return;
        }
        const numId = this.strToNumMapping.get(id) as number;
        this.strToNumMapping.delete(id);
        this.numToStrMapping.delete(numId);
    };

    public clear = () => {
        this.strToNumMapping.clear();
        this.numToStrMapping.clear();
    };
}

export { IdMapper };
